File: src/main/java/de/istkorrekt/gameoflife/ArrayMatchfield.java
package de.istkorrekt.gameoflife;

import java.util.Collection;
import java.util.function.Function;
import java.util.stream.IntStream;
import java.util.stream.Stream;

class ArrayMatchfield implements Matchfield {

    private final Cell[][] cells;

    ArrayMatchfield(Collection<Location> locations, Size size) {
        cells = emptyCells(size.getWidth(), size.getHeight());
        locations.forEach(location -> cells[location.getY()][location.getX()] = Cell.alive());
    }

    private ArrayMatchfield(Cell[][] cells) {
        this.cells = cells;
    }

    @Override
    public Cell getCellAt(Location location) {
        Cell cell = cells[location.getY()][location.getX()];
        if (cell == null) {
            return Cell.dead();
        }
        return cell;
    }

    @Override
    public Stream<Location> streamAllLocations() {
        return IntStream.range(0, width())
                .mapToObj(this::streamLocationsWithX)
                .flatMap(Function.identity());
    }

    private Stream<Location> streamLocationsWithX(int x) {
        return Stream.iterate(0, y -> y + 1)
                .map(y -> Location.of(x, y))
                .limit(height());
    }

    @Override
    public ArrayMatchfield nextGeneration() {
        Cell[][] nextGenerationCells = emptyCells(width(), height());
        streamAllLocations()
                .forEach(location -> {
                    Cell cell = getCellAt(location);
                    long amountOfLivingNeighbors = streamNeighborCells(location)
                            .filter(Cell::isAlive)
                            .count();
                    Cell nextGenerationCell = cell.nextGeneration(amountOfLivingNeighbors);
                    nextGenerationCells[location.getY()][location.getX()] = nextGenerationCell;
                });

        return new ArrayMatchfield(nextGenerationCells);
    }

    Stream<Cell> streamNeighborCells(Location location) {
        return streamNeighborLocations(location)
                .map(this::getCellAt);
    }

    Stream<Location> streamNeighborLocations(Location location) {
        int maxIndexX = maxIndexX();
        int maxIndexY = maxIndexY();
        int x = location.getX();
        int y = location.getY();
        int leftX = x > 0 ? x - 1 : maxIndexX;
        int rightX = x < maxIndexX ? x + 1 : 0;
        int topY = y > 0 ? y - 1 : maxIndexY;
        int bottomY = y < maxIndexY ? y + 1 : 0;
        return Stream.of(
                Location.of(leftX, topY),
                Location.of(x, topY),
                Location.of(rightX, topY),
                Location.of(leftX, y),
                Location.of(rightX, y),
                Location.of(leftX, bottomY),
                Location.of(x, bottomY),
                Location.of(rightX, bottomY)
        );
    }

    int width() {
        return cells[0].length;
    }

    private int maxIndexX() {
        return width() - 1;
    }

    int height() {
        return cells.length;
    }

    private int maxIndexY() {
        return height() - 1;
    }

    @Override
    public Size getSize() {
        return Size.of(width(), height());
    }

    @Override
    public String toString() {
        int maxIndexX = maxIndexX();
        int maxIndexY = maxIndexY();
        StringBuilder sb = new StringBuilder();
        for (int y = 0; y < maxIndexY; y++) {
            for (int x = 0; x < maxIndexX; x++) {
                sb.append(getCellAt(Location.of(x, y)).toString());
            }
            sb.append('\n');
        }
        return sb.toString();
    }

    @Override
    public void switchCellAt(Location location) {
        Cell existingCell = getCellAt(location);
        cells[location.getY()][location.getX()] = existingCell.isAlive()
                ? Cell.dead()
                : Cell.alive();
    }

    private static Cell[][] emptyCells(int width, int height) {
        return new Cell[height][width];
    }
}

File: src/main/kotlin/de/istkorrekt/gameoflife/ArrayMatchfield.kt
package de.istkorrekt.gameoflife

class ArrayMatchfield(private val size: Size) : Matchfield {

    private val cells: Array<Array<Cell>> = Array(size.height) { Array(size.width) { Cell.DEAD } }

    constructor(size: Size, aliveLocations: Collection<Location>) : this(size) {
        aliveLocations.forEach {
            cells[it.y][it.x] = Cell.ALIVE
        }
    }

    override fun getCellAt(location: Location): Cell {
        return cells[location.y][location.x]
    }

    override fun streamAllLocations(): Collection<Location> {
        return IntRange(0, size.width - 1)
                .flatMap { streamLocationsWithX(it) }
    }

    private fun streamLocationsWithX(x: Int): Collection<Location> {
        return IntRange(0, size.height - 1)
                .map { Location(x, it) }
    }

    override fun nextGeneration(): Matchfield {
        return ArrayMatchfield(size).also { next ->
            streamAllLocations().forEach { location ->
                val cell = getCellAt(location)
                val amountOfLivingNeighbors = collectNeighborLocations(location)
                        .map { getCellAt(it) }
                        .count { it == Cell.ALIVE }
                val nextGenerationCell = cell.nextGeneration(amountOfLivingNeighbors)
                next.setCellAt(location, nextGenerationCell)
            }
            return next
        }
    }

    fun collectNeighborLocations(location: Location): Collection<Location> {
        val maxIndexX = maxIndexX()
        val maxIndexY = maxIndexY()
        val x = location.x
        val y = location.y
        val leftX = if (x > 0) x - 1 else maxIndexX
        val rightX = if (x < maxIndexX) x + 1 else 0
        val topY = if (y > 0) y - 1 else maxIndexY
        val bottomY = if (y < maxIndexY) y + 1 else 0
        return listOf(
                Location(leftX, topY),
                Location(x, topY),
                Location(rightX, topY),
                Location(leftX, y),
                Location(rightX, y),
                Location(leftX, bottomY),
                Location(x, bottomY),
                Location(rightX, bottomY)
        )
    }

    private fun maxIndexX(): Int {
        return size.width - 1
    }

    private fun maxIndexY(): Int {
        return size.height - 1
    }

    override fun switchCellAt(location: Location) {
        cells[location.y][location.x] = when (getCellAt(location)) {
            Cell.ALIVE -> Cell.DEAD
            else -> Cell.ALIVE
        }
    }

    override fun getSize(): Size {
        return size
    }

    private fun setCellAt(location: Location, cell: Cell) {
        cells[location.y][location.x] = cell
    }

    override fun toString(): String {
        val sb = StringBuilder()
        for (y in 0 until size.height) {
            for (x in 0 until size.width) {
                sb.append(getCellAt(Location(x, y)).toString())
            }
            sb.append('\n')
        }
        return sb.toString()
    }
}